code/modules/power/supermatter/supermatter.dm

```diff
@@ -1,395 +1,588 @@
+
+		//Calculate how much gas to release
+		removed.toxins += max((device_energy * dynamic_heat_modifier) / PLASMA_RELEASE_MODIFIER, 0)
+
+		removed.oxygen += max(((device_energy + removed.temperature * dynamic_heat_modifier) - T0C) / OXYGEN_RELEASE_MODIFIER, 0)
+
+		if(produces_gas)
+			env.merge(removed)
+			air_update_turf()
+
+	for(var/mob/living/carbon/human/l in view(src, HALLUCINATION_RANGE(power))) // If they can see it without mesons on.  Bad on them.
+		if(!istype(l.glasses, /obj/item/clothing/glasses/meson))
+			var/D = sqrt(1 / max(1, get_dist(l, src)))
+			l.hallucination += power * config_hallucination_power * D
+			l.hallucination = Clamp(0, 200, l.hallucination)
 
 	for(var/mob/living/l in range(src, round((power / 100) ** 0.25)))
-		var/rads = (power / 10) * sqrt( 1 / max(get_dist(l, src),1) )
+		var/rads = (power / 10) * sqrt( 1 / max(get_dist(l, src), 1) )
 		l.apply_effect(rads, IRRADIATE)
 
-	power -= (power/DECAY_FACTOR)**3
-	handle_admin_warnings()
+	power -= ((power / 500) ** 3) * powerloss_inhibitor
+
+	if(power > POWER_PENALTY_THRESHOLD || damage > damage_penalty_point)
+
+		if(power > POWER_PENALTY_THRESHOLD)
+			playsound(loc, 'sound/weapons/emitter2.ogg', 100, 1, extrarange = 10)
+			supermatter_zap(src, 5, min(power * 2, 20000))
+			supermatter_zap(src, 5, min(power * 2, 20000))
+			if(power > SEVERE_POWER_PENALTY_THRESHOLD)
+				supermatter_zap(src, 5, min(power * 2, 20000))
+				if(power > CRITICAL_POWER_PENALTY_THRESHOLD)
+					supermatter_zap(src, 5, min(power * 2, 20000))
+		else if(damage > damage_penalty_point && prob(20))
+			playsound(loc, 'sound/weapons/emitter2.ogg', 100, 1, extrarange = 10)
+			supermatter_zap(src, 5, Clamp(power * 2, 4000, 20000))
+
+		if(prob(15) && power > POWER_PENALTY_THRESHOLD)
+			supermatter_pull(src, power / 750)
+		if(prob(5))
+			supermatter_anomaly_gen(src, FLUX_ANOMALY, rand(5, 10))
+		if(power > SEVERE_POWER_PENALTY_THRESHOLD && prob(5) || prob(1))
+			supermatter_anomaly_gen(src, GRAVITATIONAL_ANOMALY, rand(5, 10))
+		if(power > SEVERE_POWER_PENALTY_THRESHOLD && prob(2) || prob(0.3) && power > POWER_PENALTY_THRESHOLD)
+			supermatter_anomaly_gen(src, PYRO_ANOMALY, rand(5, 10))
 
-	return 1
+	if(damage > warning_point) // while the core is still damaged and it's still worth noting its status
+		if((REALTIMEOFDAY - lastwarning) / 10 >= WARNING_DELAY)
+			alarm()
+
+			if(damage > emergency_point)
+				radio.autosay("[emergency_alert] Integrity: [get_integrity()]%", name, null, list(z))
+				lastwarning = REALTIMEOFDAY
+				if(!has_reached_emergency)
+					investigate_log("has reached the emergency point for the first time.", "supermatter")
+					message_admins("[src] has reached the emergency point [ADMIN_JMP(src)].")
+					has_reached_emergency = TRUE
+			else if(damage >= damage_archived) // The damage is still going up
+				radio.autosay("[warning_alert] Integrity: [get_integrity()]%", name, "Engineering", list(z))
+				lastwarning = REALTIMEOFDAY - (WARNING_DELAY * 5)
```

Comment:
Why the `* 5` here? Just a minus will also make it warn engineering every processing tick